flaml/tune/space.py

```diff
@@ -76,3 +83,74 @@ def define_by_run_func(
                     type(domain.sampler).__name__))
     # Return all constants in a dictionary.
     return config
+
+
+def exclusive_to_inclusive(space: Dict) -> Dict:
+    """Change the upper bound from exclusive to inclusive.
+
+    Returns:
+        A copied dict with modified upper bound.
+    """
+    space = space.copy()
+    for key in space:
+        domain = space[key]
+        if not isinstance(domain, sample.Domain):
+            continue
+        sampler = domain.get_sampler()
+        if isinstance(sampler, sample.Quantized):
+            continue
+        if isinstance(domain, sample.Integer):
+            if isinstance(sampler, sample.LogUniform):
+                space[key] = sample.lograndint(
+                    domain.lower, domain.upper - 1, sampler.base)
+            elif isinstance(sampler, sample.Uniform):
+                space[key] = sample.randint(
+                    domain.lower, domain.upper - 1)
+    return space
+
+
+def add_cost_to_space(space: Dict, low_cost_point: Dict, choice_cost: Dict):
+    """Update the space in place by adding low_cost_point and choice_cost
+    """
+    for key in space:
+        domain = space[key]
+        if not isinstance(domain, sample.Domain):
+            continue
+        low_cost = low_cost_point.get(key)
+        choice_cost_list = choice_cost.get(key)
+        if isinstance(domain, sample.Categorical):
+            for i, cat in enumerate(domain.categories):
+                if isinstance(cat, dict):
+                    if isinstance(low_cost, list):
+                        low_cost_dict = low_cost[i]
+                    else:
+                        low_cost_dict = {}
+                    if choice_cost_list:
+                        choice_cost_dict = choice_cost_list[i]
+                    else:
+                        choice_cost_dict = {}
+                    add_cost_to_space(cat, low_cost_dict, choice_cost_dict)
+            if choice_cost_list:
+                if len(choice_cost_list) == len(domain.categories):
+                    domain.choice_cost = choice_cost_list

```

Comment:
does ```sample.Categorical``` have this attribute ```choice_cost ```?  Didn't find it.

Reply:
No, I added it.

Reply:
Didn't find it. Where do you add?

Reply:
It doesn't. I added it.